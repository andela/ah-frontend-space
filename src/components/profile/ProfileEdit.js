/* eslint-disable react/no-this-in-sfc */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import profileImage from '../../assets/images/profile.png';

function buttons(onSubmit, buttonLoading) {
  return (
    <Fragment>
      <div className="buttons">

        <button type="submit" onClick={onSubmit} disabled={buttonLoading} className="btn btn-primary btn-lg">
          {buttonLoading && (<i className="spinner-grow spinner-grow-sm" />)}
          {buttonLoading && <span>Just A sec Please!...</span>}
          {!buttonLoading && <span>Update Profile</span>}
        </button>
        <Link className="btn btn-danger btn-lg cancel" to="/profile">
                Cancel
        </Link>
      </div>
    </Fragment>
  );
}


function inputs(Label, type, name, handleChange, value) {
  return (
    <Fragment>
      <p>
        {Label}
        <input style={{ 'margin-left': '30px' }} type={type} name={name} onChange={handleChange} value={value} />
      </p>
    </Fragment>
  );
}
function ProfileEditComponent({
  firstName, lastName, dateOfBirth, bio, image, onSubmit,
  handleChange, buttonLoading, imageOnChange, status,
}) {
  return (
    <Fragment>
      <div className="container-fluid">
        <hr className="hr-primary" />
        <h1 className="profiletitle">Edit Your Personal Details</h1>
        <hr className="hr-primary" />
        <form onSubmit={onSubmit}>
          <div style={{ display: 'flex' }}>
            <div style={{
              display: 'flex', flex: 0.7, flexDirection: 'column', margin: '0 auto',
            }}
            >
              {(image !== 'url') ? (
                <img
                  src={image}
                  alt=""
                  width="370em"
                  height="310em"
                  radius="6px"
                />
              ) : (
                <img
                  alt="profileimage"
                  className="default-view"
                  src={profileImage}
                  width="370em"
                  height="310em"
                  radius="6px"
                />
              )}
              <br />
              <input type="file" name={image} onChange={imageOnChange} />
              <progress value={status} max="100" />
            </div>
            <div
              className="edit-side form-control"
              style={{
                display: 'flex', flex: 1, flexDirection: 'column', margin: '0 auto',
              }}
            >
              {inputs('firstName', 'text', 'firstName', handleChange, firstName)}
              {inputs('lastName', 'text', 'lastName', handleChange, lastName)}
              {inputs('Birth Day', 'date', 'dateOfBirth', handleChange, dateOfBirth)}
              <p>
                Bio
                <textarea style={{ 'margin-left': '30px' }} type="text" value={bio} name="bio" onChange={handleChange} />
              </p>
              <br />
              {buttons(onSubmit, buttonLoading)}
            </div>
          </div>
        </form>
      </div>
    </Fragment>
  );
}


const Props = {
  firstName: PropTypes.string.isRequired,
  lastName: PropTypes.string.isRequired,
  dateOfBirth: PropTypes.string.isRequired,
  bio: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  onSubmit: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  buttonLoading: PropTypes.bool.isRequired,
  imageOnChange: PropTypes.func.isRequired,
  status: PropTypes.number.isRequired,
};

ProfileEditComponent.propTypes = Props;

export default ProfileEditComponent;
