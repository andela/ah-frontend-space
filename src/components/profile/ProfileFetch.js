/* eslint-disable camelcase */
import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import profileImage from '../../assets/images/profile.png';


function SubContent(profile, className, label) {
  return (
    <Fragment>
      <p className={className}>
        {' '}
        <strong style={{ 'font-weight': 'lighter', 'margin-right': '30px' }}>
          {label}
        </strong>
        {profile}
        {' '}
      </p>
    </Fragment>
  );
}

function ProfileComponentContent(profile) {
  console.log(profile.image, 'great');
  return (
    <Fragment>
      <div className="container-fluid">
        <div className="profile">
          <hr className="hr-primary" />
          <h1 className="profiletitle">User Profile</h1>
          <hr className="hr-primary" />
          <div className="profile-data" style={{ display: 'flex' }}>
            <div
              className="profile-side"
              style={{
                display: 'flex', flex: 1, 'flex-direction': 'column', margin: '0 auto',
              }}
            >
              {(profile.image !== 'url') ? (
                <img
                  src={profile.image}
                  alt=""
                  width="370em"
                  height="310em"
                />
              ) : (
                <img
                  alt="profileimage"
                  className="default-view"
                  src={profileImage}
                  width="370em"
                  height="310em"
                />
              )}
              <br />
            </div>
            <div className="profile-desc" style={{ display: 'flex', flex: 1.5, 'flex-direction': 'column' }}>
              <div style={{ display: 'block' }}>
                {SubContent((profile.first_name), 'firstname', 'First Name')}
                <hr className="hr-primary" />
                { SubContent((profile.last_name), 'lastname', 'Last Name')}
                <hr className="hr-primary" />
                {SubContent((profile.username), '', 'Username')}
                <hr className="hr-primary" />
                {SubContent((profile.following), '', 'Following')}
                {profile.following}
                <hr className="hr-primary" />
                {SubContent((profile.date_of_birth), '', 'Date Of Birth')}
                <hr className="hr-primary" />
                <br />
                { SubContent((profile.bio), '', 'Biography')}
                <hr className="hr-primary" />
                <br />
                <Link className="btn btn-primary edit-button" to="/editprofile">Edit Profile</Link>
              </div>
            </div>
          </div>
        </div>
      </div>
    </Fragment>
  );
}


function ProfileComponent(props) {
  const { profile: { profile } } = props;
  return (
    <div>
      {ProfileComponentContent(profile)}
    </div>
  );
}

ProfileComponent.propTypes = {
  profile: PropTypes.shape().isRequired,
};

export default ProfileComponent;
