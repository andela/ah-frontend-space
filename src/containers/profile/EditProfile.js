
/* eslint-disable no-shadow */
/* eslint-disable camelcase */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropType from 'prop-types';
import { editUserProfile, fetchProfile } from '../../actions/profileActions';
import firebase from '../../firebase/firebase.config';
import ProfileEditComponent from '../../components/profile/ProfileEdit';
import { Authenticated } from '../../routes/Authenticated';


export class ProfileIsEdited extends Component {
  constructor(props) {
    super(props);
    this.state = {
      firstName: '',
      lastName: '',
      dateOfBirth: '',
      bio: '',
      image: '',
      buttonLoading: false,
      status: 0,
    };
  }

  componentDidMount() {
    const { fetchProfile, profile: { profile } } = this.props;
    fetchProfile();
  }


  componentWillReceiveProps(nextProps) {
    const {
      first_name, last_name, username, date_of_birth, bio, image,
    } = nextProps.profile.profile;
    if (username) {
      this.setState({
        firstName: first_name,
        lastName: last_name,
        username,
        image,
        dateOfBirth: date_of_birth,
        bio,
      });
    }
  }

  fetchData = () => {
    this.setState({ buttonLoading: true });
  };

  handleChange = (event) => {
    event.preventDefault();
    this.setState({ [event.target.name]: event.target.value });
  }

  handleSubmit = (event) => {
    event.preventDefault();
    this.fetchData();
    const {
      firstName, lastName, username, dateOfBirth, bio, image,
    } = this.state;

    const ProfileData = {
      first_name: firstName,
      last_name: lastName,
      username,
      date_of_birth: dateOfBirth,
      bio,
      image,
    };
    const { editUserProfile } = this.props;
    editUserProfile(ProfileData, this.props);
  }

  /* istanbul ignore next */
  uploadImage(files) {
    const fileload = firebase
      .storage()
      .ref(`images/${files[0].name}`)
      .put(files[0]);
    fileload.then(() => {
      firebase
        .storage()
        .ref(`images/${files[0].name}`)
        .getDownloadURL()
        .then((url) => {
          const image = {
            image: url,
          };
          this.setState({ image: image.image });
        });
    });
    fileload.on('state_changed', (snapshot) => {
      const status = Math.round(
        (snapshot.bytesTransferred / snapshot.totalBytes) * 100,
      );
      this.setState({ status });
    });
  }

  render() {
    const {
      firstName, lastName, dateOfBirth, bio, image, buttonLoading, status,
    } = this.state;
    return (
      <ProfileEditComponent
        firstName={firstName}
        lastName={lastName}
        dateOfBirth={dateOfBirth}
        bio={bio}
        image={image}
        onSubmit={this.handleSubmit}
        handleChange={this.handleChange}
        buttonLoading={buttonLoading}
        uploadImage={this.uploadImage}
        imageOnChange={event => this.uploadImage(event.target.files)}
        status={status}

      />
    );
  }
}

export const mapStateToProps = state => ({
  profile: state.profile,
});

ProfileIsEdited.defaultProps = {
  fetchProfile: () => {},
};

ProfileIsEdited.propTypes = {
  fetchProfile: PropType.func,
}.isRequired;

export default connect(
  mapStateToProps,
  { fetchProfile, editUserProfile },
)(Authenticated(ProfileIsEdited));
