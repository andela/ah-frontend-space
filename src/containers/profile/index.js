/* eslint-disable react/prop-types */
/* eslint-disable no-shadow */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropType from 'prop-types';
import { fetchProfile } from '../../actions/profileActions';
import ProfileComponent from '../../components/profile/ProfileFetch';
import { Authenticated } from '../../routes/Authenticated';


export class Profile extends Component {
  componentDidMount() {
    const { fetchProfile } = this.props;
    fetchProfile();
  }

  render() {
    const { profile } = this.props;
    return (<ProfileComponent profile={profile} />);
  }
}

export const mapStateToProps = state => (
  {
    profile: state.profile,
  }
);

Profile.propType = {
  props: PropType.object.isRequired,
  fetchProfile: PropType.func.isRequired,
  profile: PropType.object.isRequired,
};

export default connect(mapStateToProps, { fetchProfile })(Authenticated(Profile));
