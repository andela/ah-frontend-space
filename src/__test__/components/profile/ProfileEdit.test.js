import React from 'react';
import { shallow } from 'enzyme';
import { ProfileIsEdited, mapStateToProps } from '../../../containers/profile/EditProfile';


describe('ProfileIsEdited', () => {
  let wrapper;
  const props = {
    profileIsEdited: jest.fn(),
    profile: {
      profile: {
        bio: 'cool',
        firstName: 'first_name',
      },
    },

    fetchProfile: jest.fn(),
    editUserProfile: jest.fn(),
    componentWillReceiveProps: jest.fn(),
  };
  beforeEach(() => {
    wrapper = shallow(<ProfileIsEdited {...props} />);
  });
  it('renders without fail', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('it should map state to props', () => {
    const initialState = {
      profile: '',
    };
    expect(mapStateToProps(initialState)).toEqual({
      profile: '',
    });
  });

  it('should detect change in state and dispatch the required action', () => {
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'first_name',
        value: 'first',
      },
    };
    const wrapperInst = wrapper.instance();
    wrapperInst.handleChange(event);
    wrapperInst.handleSubmit(event);
    expect(wrapperInst.state.first_name).toBe('first');
  });

  it('should set new state for props', () => {
    const nextProps = {
      profile: {
        profile: {
          username: 'bxbhdhhsaj',
          first_name: 'firstname',
          last_name: 'last',
          bio: 'bio',
          image: 'http//',
          date_of_birth: '2001-02-02',
        },
      },

    };
    wrapper.setProps({ ...nextProps });
    expect(nextProps.profile.profile.first_name).toEqual('firstname');
  });
});
