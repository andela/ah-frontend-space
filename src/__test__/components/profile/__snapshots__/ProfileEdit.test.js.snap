// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileIsEdited renders without fail 1`] = `ShallowWrapper {}`;
